### PR TITLE
[SPARK-36519][SS]Store RocksDB format version in the checkpoint for streaming queries

### PR DESCRIPTION
**Pull request by zsxwing**
### What changes were proposed in this pull request?

RocksDB provides backward compatibility but it doesn't always provide forward compatibility. It's better to store the RocksDB format version in the checkpoint so that it would give us more information to provide the rollback guarantee when we upgrade the RocksDB version that may introduce incompatible change in a new Spark version.

A typical case is when a user upgrades their query to a new Spark version, and this new Spark version has a new RocksDB version which may use a new format. But the user hits some bug and decide to rollback. But in the old Spark version, the old RocksDB version cannot read the new format.

In order to handle this case, we will write the RocksDB format version to the checkpoint. When restarting from a checkpoint, we will force RocksDB to use the format version stored in the checkpoint. This will ensure the user can rollback their Spark version if needed.

We also provide a config `spark.sql.streaming.stateStore.rocksdb.formatVersion` for users who don't need to rollback their Spark versions to overwrite the format version specified in the checkpoint.

### Why are the changes needed?

Provide the Spark version rollback guarantee for streaming queries when a new RocksDB introduces an incompatible format change.

### Does this PR introduce _any_ user-facing change?

No. RocksDB state store is a new feature in Spark 3.2, which has not yet released.

### How was this patch tested?

The new unit tests.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala:65]: Changed this test to show the bug of `ConfEntry.fullName`: SQLConf is always case sensitive but currently `ConfEntry.fullName` returns an incorrect lower case name in the error message.

This test will fail if we add `toLowerCase` back to `ConfEntry.fullName`.

**Comment by zsxwing** [comment]
@xuanyuanking @HeartSaVioR

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:518]: Nice explanation! It would be nice if we can refer this from config in SQLConf which is closer to user facing - despite it's marked as internal, they find the config in SQLConf first instead of this.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala:65]: Should we remove this as well in RocksDBConf if we want to have consistent behavior, "case sensitive"?
`val confs = CaseInsensitiveMap[String](storeConf.confs)`

**Comment by HeartSaVioR** [comment]
cc. @viirya

**Comment by HeartSaVioR** [comment]
cc. @gengliangwang as well as this PR targets to Spark 3.2.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1602]: +1, it's a little strange that the version starts from 5.

**Review by xuanyuanking — APPROVED** [review]
LGTM
Just some nits.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:523]: Once we use the format version in the checkpoint when restarting a query, it will just use original format on the RocksDB instead of the new format, why needing to rollback Spark version?

**Review by viirya — COMMENTED** [review]
Overall looks good. Just wondering if an optional config is better and a question about the comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala
Comment (zsxwing) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:519]: > Nice explanation! It would be nice if we can refer this from config in SQLConf which is closer to user facing - despite it's marked as internal, they find the config in SQLConf first instead of this.

Made some improvement for the doc of `SQLConf.STATE_STORE_ROCKSDB_FORMAT_VERSION`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala
Comment (zsxwing) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:524]: This is to encourage users to upgrade  their Spark. If we tell a user that if they upgrade the Spark version for an existing production query, they could not rollback to the old Spark version because old Spark version cannot read the new checkpoint format, they probably would hesitate to upgrade Spark (They may hit regressions in new Spark version).

**Comment by zsxwing** [comment]
I reverted the changes for the `fullName` bug. Feel it's better to fix it in a separate PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (zsxwing) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1597]: We would like to write `5` to the checkpoint when the config is not set. That's why we need a default value.

**Comment by HeartSaVioR** [comment]
I see the benefit of ability on overriding the table format version, but it also brings confusion as well. This is not the same as we have been doing from state format version.

For state format version, we restrict modifying the value once it is set in any way (regardless default value or user defined value), so the value is unchanged in the lifetime of the query. Here we allow to set any value to override the config, which leads that multiple table format versions exist in SST files. This won't hurt if RocksDB version in Spark can understand all of these versions, but it is still less simpler.

If we'd like to make this simpler as possible, it'd be better if we just disallow overriding the config. If we still want to open the chance to override based on the behavior that having multiple table format versions doesn't matter, let's restrict setting "lower" version than checkpoint to ensure there are no SST file having higher table format version than checkpoint config.

**Comment by gengliangwang** [comment]
@zsxwing @HeartSaVioR I am going to cut RC1 this week. Please come to an agreement on the version overriding ASAP. Thanks!

**Comment by HeartSaVioR** [comment]
I'm OK in either way I proposed. I'd like to see the approach be changed as there's a possibility where SST file has higher version than the version marked in checkpoint, and it makes the problem uneasy to discover and resolve.

**Comment by zsxwing** [comment]
@HeartSaVioR Yeah, totally agree that this is too complicated. Since this is the first release of RocksDB state store, allowing users to overwrite the config value in the checkpoint is not needed. We can revisit this in future when RocksDB has a new format. I updated the PR to disallow overwriting. So a query will use the same rocksdb format version during its entire lifttime.

**Review by HeartSaVioR — APPROVED** [review]
+1 Thanks for the quick fix!

**Comment by HeartSaVioR** [comment]
I'll merge this early tomorrow if there's no further comment, or @viirya is OK with this. cc. @viirya

**Comment by gengliangwang** [comment]
I will cut 3.2.0 RC1 after this one is merged. 
cc @viirya

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Thanks! Merging to master/3.2.

**Pull request closed without merging**
